parallel-install/pkg/preinstaller/resourceapplier.go

```diff
@@ -0,0 +1,124 @@
+package preinstaller
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/ghodss/yaml"
+	"github.com/kyma-incubator/hydroform/parallel-install/pkg/logger"
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"strings"
+)
+
+// ResourceApplier creates a new resource from manifest on k8s cluster.
+type ResourceApplier interface {
+	// Apply parses passed manifest and applies it on a k8s cluster.
+	Apply(manifest string) (bool, error)
+}
+
+type resourceType struct {
+	name string
+}
+
+// GenericResourceApplier is a default implementation of ResourceApplier.
+type GenericResourceApplier struct {
+	log             logger.Interface
+	decoder         runtime.Decoder
+	resourceManager ResourceManager
+}
+
+// NewGenericResourceApplier returns a new instance of GenericResourceApplier.
+func NewGenericResourceApplier(log logger.Interface, resourceManager ResourceManager) *GenericResourceApplier {
+	return &GenericResourceApplier{
+		log:             log,
+		decoder:         initializeDecoder(),
+		resourceManager: resourceManager,
+	}
+}
+
+func (c *GenericResourceApplier) Apply(manifest string) (bool, error) {
```

Comment:
I am missing a clear contract for this method.
It returns "false, nil" in four cases. Three of these are related to parsing/decoding input files. The fourth one is when the resource already exists... This is very dangerous for upgrades. We may silently proceed with an upgrade even if one CRD was not updated to the latest version because it already exists in the cluster.
I would say we should either:
- decide it's an error when resource already exist
- return a specific, unique value that allows the caller to recognize the "resource already exists" condition and leave the decision "what to do next" to the caller.

Reply:
Agree. Contract has been changed. Now it will return error only, so fail-fast approach will be taken.